### DISCUSSION TITLE
"As we can add only one sensor method..." (webserver)

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#21221] [Q&A]
After mentally differentiating between drivers and modules, I now see the WebServer example in a new light. Previously, I assumed that I had to organize all WebUI extensions in one file or module. Now I understand that practically every class could be a driver, and thus I can implement the WebUI extension within the classes.

In the [example ](https://tasmota.github.io/docs/Berry-Cookbook/#adding-a-button-to-the-main-menu):
![grafik](https://github.com/arendst/Tasmota/assets/16272493/2b6a9573-38ff-4dc7-8871-2eabf0218361)
Does this mean that there can only be one web_sensor() per driver, or only one in general? If second, I would need to gather all 'has_arg()' queries across them, which doesn't make much sense to me.

**Comment by sfromis** [comment]
One per driver, in many drivers. The base purpose of `web_sensor` does not include using `webserver.has_arg`, as the aim is that it can use `tasmota.web_send` or `tasmota.web_send_decimal` to provide data for the web UI front page. And Tasmota can have a long list of sensors there, Berry or not.

I've not tried using `webserver` module methods in `web_sensor` of multiple drivers, but do not see any issue with using `has_arg` etc, as long as you have separate arguments per driver.

I was a bit surprised to see two vastly different usages using the same callback method `web_sensor`, but that seems to be how it is.

The cookbook example alludes to that you can have one driver with multiple web UI elements, but the same `web_sensor` method has to deal with all of them, instead of being able to have one callback per element.

**Reply by ChrisVonFrosta** [reply]
>One per driver, in many drivers. The base purpose of `web_sensor`does not include using `webserver.has_arg`, as the aim is that it can use `tasmota.web_send` or `tasmota.web_send_decimal` to provide data for the web UI front page. And Tasmota can have a long list of sensors there, Berry or not.

thx. even if `webserver.has_arg` is not the base purpose of `web_sensor` that is the right place for it, isn't it? 
I think you say that here:
>I was a bit surprised to see two vastly different usages using the same callback method `web_sensor`, but that seems to be how it is.

as you point on `tasmot.web_send`: can the counter value be hidden in web UI? I want to calculate the counter to other value.

**Reply by sfromis** [reply]
Right .. and no.
